core/src/pcap.rs

```diff
@@ -0,0 +1,226 @@
+        }
+
+        Ok(())
+    }
+
+    fn flush(&self) -> Result<()> {
+        unsafe {
+            ffi::pcap_dump_flush(self.dumper.as_ptr())
+                .to_result()
+                .map(|_| ())
+        }
+    }
+}
+
+impl Drop for Pcap {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::pcap_dump_close(self.dumper.as_ptr());
+            ffi::pcap_close(self.handle.as_ptr());
+        }
+    }
+}
+
+/// Generate PCAP files for rx/tx queues per port and per core.
+pub(crate) fn create_for_queues(port: PortId, core: CoreId) -> Result<()> {
+    Pcap::create(format!("{:?}-{:?}-rx.pcap", port, core).as_str())?;
+    Pcap::create(format!("{:?}-{:?}-tx.pcap", port, core).as_str())?;
+    Ok(())
+}
+
+/// Append and write slice of packets to an already-created file, logging errors
+/// that may occur.
+pub(crate) fn append_and_write<T: Packet>(
+    port: PortId,
+    core: CoreId,
+    tx_or_rx: &str,
+    packets: &[T],
+) {
+    let path = String::from(format!("{:?}-{:?}-{}.pcap", port, core, tx_or_rx).as_str());
+    let _ = Pcap::append(path.as_str())
+        .map_err(|err| {
+            error!(
+                message = "can't append to pcap file",
+                pcap = path.as_str(),
+                ?err
+            )
+        })
+        .map(|pcap| {
+            pcap.write(&packets).map_err(|err| {
+                error!(
+                    message = "can't write packets to pcap file",
+                    pcap = path.as_str(),
+                    ?err
+                )
+            })
+        });
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::packets::ip::v4::Ipv4;
+    use crate::packets::{Ethernet, Udp, UDP_PACKET};
+    use crate::Mbuf;
+    use std::fs;
+    use std::ptr;
+
+    fn read_pcap_plen(path: &str) -> u32 {
+        let mut errbuf = [0i8; ffi::RTE_MBUF_DEFAULT_BUF_SIZE as usize];
+        let handle =
+            unsafe { ffi::pcap_open_offline(path.to_cstring().as_ptr(), errbuf.as_mut_ptr()) };
+
+        let mut header: *mut ffi::pcap_pkthdr = ptr::null_mut();
+        let mut buf: *const libc::c_uchar = ptr::null();
+
+        let mut ret = 0;
+
+        while let 1 = unsafe { ffi::pcap_next_ex(handle, &mut header, &mut buf) } {
+            ret += unsafe { (*header).caplen }
+        }
+
+        ret
+    }
+
+    fn cleanup(path: &str) {
+        fs::remove_file(path).unwrap();
+    }
+
+    #[nb2::test]
+    fn create_pcap_and_write_packet() {
+        let writer = Pcap::create("foo.pcap").unwrap();
+        let packet = Mbuf::from_bytes(&UDP_PACKET).unwrap();
+        let data_len = packet.data_len();
+        let ethernet = packet.parse::<Ethernet>().unwrap();
+        let ipv4 = ethernet.parse::<Ipv4>().unwrap();
+        let udp = ipv4.parse::<Udp<Ipv4>>().unwrap();
+
+        let res = writer.write(&[udp]);
+
+        assert!(res.is_ok());
+        let len = read_pcap_plen("foo.pcap");
+        assert_eq!(data_len as u32, len);
+        cleanup("foo.pcap");
```

Comment:
only thing here is if the assertion fails, the cleanup won't happen due to panic. but I think that's fine. `create` in the next run will start a new file. so it's a best effort cleanup.

Reply:
Yep and w e gitignore pcaps.